hack/verify-code-patterns.sh

```diff
@@ -43,3 +43,17 @@ if [[ -n "$out" ]]; then
   echo >&2 "$out"
   exit 1
 fi
+
+# Do not initialize index.{Plugin,Platform} structs in test code.
+#
+# TODO(ahmetb): ideally this plugin should cleanly scan all (index\.)(Plugin|Platform) occurrences except slice literals
+# (which are prefixed with []). However I could not get negative lookbehind pattern to work with grep to match to ones
+# that do not have a [] prefix.
+out="$(grep --include '*_test.go' --exclude-dir 'vendor/' -EIrn '\s+(index\.)(Plugin|Platform){' || true)"
```

Comment:
```suggestion
out="$(grep --include '*_test.go' --exclude-dir 'vendor/' -EIrn '[^]](index\.)(Plugin|Platform){' || true)"
```
This works with GNU grep. Can you check?

Reply:
Can you explain `[^]]`? To me it reads like `[^]` is a negated set but it's empty, then there's a `]` literal outside that's supposed to match.

Reply:
I didn't know either. But if there is a negated character set, the first char following the caret is always interpreted verbatim. So this is indeed a negated char set of just `]`. Iow it matches anything but a `]`, which is what we need here, right?

Reply:
try on https://www.regexr.com

Reply:
https://regexr.com/4ht16

Reply:
I didn't find out what regex standard regexr.com uses. My GNU grep 3.1 does what I described, I just double-checked. Have you checked, if it also works with BSD grep?

Reply:
I finally found a proper reference: `man 7 regex` says

> To  include a literal ']' in the list, make it the first character (following a possible '^').  To include a literal '-', make it the first or last character, or the
       second endpoint of a range.

Reply:
For gnu grep, `[^]]` and `[^\]]` are different things:

```
$ grep --include '*_test.go' --exclude-dir 'vendor/' -EIrn '[^\]](index\.)(Plugin|Platform){'
cmd/validate-krew-manifest/main_test.go:177:	err := isOverlappingPlatformSelectors([]index.Platform{p1, p2})
cmd/validate-krew-manifest/main_test.go:192:	err := isOverlappingPlatformSelectors([]index.Platform{p1, p2})

$ grep --include '*_test.go' --exclude-dir 'vendor/' -EIrn '[^]](index\.)(Plugin|Platform){'
cmd/validate-krew-manifest/main_test.go:169:	p1 := index.Platform{
cmd/validate-krew-manifest/main_test.go:174:	p2 := index.Platform{
cmd/validate-krew-manifest/main_test.go:184:	p1 := index.Platform{
cmd/validate-krew-manifest/main_test.go:189:	p2 := index.Platform{
```

BSD grep is **very** dumb. It accepts arguments but then accepts input from STDIN (its cli is not fully compatible with gnu grep).

```
$ /usr/bin/grep --include '*_test.go' --exclude-dir 'vendor/' -EIrn '[^]](index\.)(Plugin|Platform){' | cat
grep: warning: recursive search of stdin
```

maybe we should try to fix but I don't care much.

Reply:
I've incorporated your recommendation.